### PR TITLE
task 10, 20 , 30

### PR DESCRIPTION
**Pull request by KonstantinLevchukGree**

### CODE REVIEW

File: src/test/resources/clientData.properties [hunk 1]
```diff
@@ -0,0 +1,9 @@
+writer.user.name=0oa157tvtugfFXEhU4x7
+writer.user.password=X7eBCXqlFC7x-mjxG5H91IRv_Bqe1oq7ZwXNA8aq
+writer.user.grant.type=client_credentials
+writer.user.scope=write
+
+reader.user.name=0oa157tvtugfFXEhU4x7
```
Comment (KirKarpov) [inline src/test/resources/clientData.properties]: api.zip.code=${base.host}/zipcode
api.expand=${api.zip.code}/expands
localhost:9090/zipcodes/expand

File: src/main/java/user/User.java [hunk 2]
```diff
@@ -0,0 +1,67 @@
+package user;
+
+import com.fasterxml.jackson.annotation.JsonIgnoreProperties;
+
+import java.util.Objects;
+@JsonIgnoreProperties(ignoreUnknown = true)
+
```
Comment (KirKarpov) [inline src/main/java/user/User.java]: No need for empty line

File: src/main/java/user/User.java [hunk 3]
```diff
@@ -0,0 +1,67 @@
+package user;
+
+import com.fasterxml.jackson.annotation.JsonIgnoreProperties;
+
+import java.util.Objects;
+@JsonIgnoreProperties(ignoreUnknown = true)
```
Comment (KirKarpov) [inline src/main/java/user/User.java]: Please add empty line

File: src/main/java/token/Token.java [hunk 4]
```diff
@@ -0,0 +1,12 @@
...
+
+import com.fasterxml.jackson.annotation.JsonIgnoreProperties;
+
+@JsonIgnoreProperties(ignoreUnknown = true)
+public class Token {
+    public String access_token;
+
+    public String getAccess_token() {
+        return access_token;
+    }
```
Comment (KirKarpov) [inline src/main/java/token/Token.java]: You can use Lombook annotation

File: src/main/java/utils/api/ApiUtils.java [hunk 5]
```diff
@@ -0,0 +1,111 @@
...
+        CloseableHttpResponse response = httpclient.execute(httpget);
+        int statusCode = response.getStatusLine().getStatusCode();
+        assertThat(statusCode, equalTo(statusCod));
+        String jsonString;
+        try {
+            jsonString = EntityUtils.toString(response.getEntity());
+        } finally {
+            response.close();
+        }
+        return jsonString;
```
Comment (KirKarpov) [inline src/main/java/utils/api/ApiUtils.java]: Should be moved to the method the works with response. It should be moved to Utils

File: src/main/java/utils/api/ApiUtils.java [hunk 6]
```diff
@@ -0,0 +1,111 @@
...
+
+    @SneakyThrows
+    private static URI createApiMethodUri(String token, String methodPoint, User user) {
+        return new URIBuilder(new URI(Point.BASE_URL.getApiPoint() + methodPoint))
+                .setParameter("access_token", token)
+                .setParameter("olderThan", String.valueOf(user.getAge() - 1))
+                .setParameter("sex", user.getSex())
+                .build();
+    }
+}
```
Comment (KirKarpov) [inline src/main/java/utils/api/ApiUtils.java]: Should be moved to Utils

File: src/main/java/utils/api/ApiUtils.java [hunk 7]
```diff
@@ -0,0 +1,111 @@
...
+    }
+
+    @SneakyThrows
+    private static URI createApiMethodUri(String token, String methodPoint, User user) {
+        return new URIBuilder(new URI(Point.BASE_URL.getApiPoint() + methodPoint))
+                .setParameter("access_token", token)
+                .setParameter("olderThan", String.valueOf(user.getAge() - 1))
+                .setParameter("sex", user.getSex())
+                .build();
+    }
```
Comment (KirKarpov) [inline src/main/java/utils/api/ApiUtils.java]: I am not sure that you need those methods

File: src/main/java/utils/authenticationClient/AuthenticationClient.java [hunk 8]
```diff
@@ -0,0 +1,75 @@
...
+        return new HttpPost(Point.BASE_URL.getApiPoint() + Point.USER_TOKEN.getApiPoint());
+    }
+
+    private static List<NameValuePair> putParametersPost(String grantTypeName, String scopeName) {
+        List<NameValuePair> nvp = new ArrayList<>();
+        nvp.add(new BasicNameValuePair("grant_type", grantTypeName));
+        nvp.add(new BasicNameValuePair("scope", scopeName));
+        return nvp;
+    }
+
```
Comment (KirKarpov) [inline src/main/java/utils/authenticationClient/AuthenticationClient.java]: Those methods complicate code reading. Could you please remove them and write the code inside the methods where it needed.

File: src/test/java/tests/BaseTest.java [hunk 9]
```diff
@@ -0,0 +1,46 @@
...
+    private  final Properties clientData = PropertyUtil.getProperties("clientData.properties");
+    private  final Properties httpHost = PropertyUtil.getProperties("httpHost.properties");
+    private  final Properties testData = PropertyUtil.getProperties("testData.properties");
+    @BeforeEach
+    public void init() {
+        writerClient = new Client();
+        writerClient.setUserName(clientData.getProperty("writer.user.name"));
+        writerClient.setUserPassword(clientData.getProperty("writer.user.password"));
+        writerClient.setGrantType(clientData.getProperty("writer.user.grant.type"));
+        writerClient.setUserScope(clientData.getProperty("writer.user.scope"));
```
Comment (KirKarpov) [inline src/test/java/tests/BaseTest.java]: Please create constructor with parameters and create object via constructor

File: src/test/java/tests/BaseTest.java [hunk 10]
```diff
@@ -0,0 +1,46 @@
...
+
+        readerClient = new Client();
+        readerClient.setUserName(clientData.getProperty("reader.user.name"));
+        readerClient.setUserPassword(clientData.getProperty("reader.user.password"));
+        readerClient.setGrantType(clientData.getProperty("reader.user.grant.type"));
+        readerClient.setUserScope(clientData.getProperty("reader.user.scope"));
```
Comment (KirKarpov) [inline src/test/java/tests/BaseTest.java]: same there

File: src/test/java/tests/BaseTest.java [hunk 11]
```diff
@@ -0,0 +1,46 @@
...
+
+        readerToken = AuthenticationClient.getToken(readerClient,httpHost.getProperty("port"),httpHost.getProperty("host.name"),httpHost.getProperty("scheme"));
+        writerToken = AuthenticationClient.getToken(writerClient,httpHost.getProperty("port"),httpHost.getProperty("host.name"),httpHost.getProperty("scheme"));
+
+        user=new User();
+        user.setName(testData.getProperty("user.name"));
+        user.setAge(Integer.parseInt(testData.getProperty("user.age")));
+        user.setSex(testData.getProperty("user.sex"));
```
Comment (KirKarpov) [inline src/test/java/tests/BaseTest.java]: same there

File: src/test/java/tests/UserTest.java [hunk 12]
```diff
@@ -0,0 +1,64 @@
...
+
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.junit.jupiter.api.Assertions.assertFalse;
+
+public class UserTest extends BaseTest {
+
+    @Test
+    public void checkAddUserWithAllFields() {
+        String zipCodesJson = ApiUtils.sendGet(readerToken, Point.ZIP_CODES.getApiPoint(), HttpStatus.SC_CREATED);
+        List<String> zipCodesList = ApiUtils.getListFromJson(zipCodesJson);
```
Comment (KirKarpov) [inline src/test/java/tests/UserTest.java]: getZipCodes ()
return List<String>

File: src/test/java/tests/UserTest.java [hunk 13]
```diff
@@ -0,0 +1,64 @@
...
+        String zipCod = zipCodesList.get(0);
```
Comment (KirKarpov) [inline src/test/java/tests/UserTest.java]: String zipCod = getZipCodes().get(0);

File: src/test/java/tests/UserTest.java [hunk 14]
```diff
@@ -0,0 +1,64 @@
...
+        user.setZipCode(zipCod);
+        String userJson = ApiUtils.getJsonStringFromObject(user);
+        ApiUtils.sendPost(writerToken, Point.USERS.getApiPoint(), userJson, HttpStatus.SC_CREATED);
```
Comment (KirKarpov) [inline src/test/java/tests/UserTest.java]: write(send)User(zipCode);

File: src/test/java/tests/UserTest.java [hunk 15]
```diff
@@ -0,0 +1,64 @@
...
+        String userJsonFromApi = ApiUtils.sendGet(readerToken, Point.USERS.getApiPoint(), HttpStatus.SC_OK, user)
+                .replace("[", "").replace("]", "").trim();
+        List<User> userFromApi = ApiUtils.getUserFromJsonString(userJsonFromApi);
```
Comment (KirKarpov) [inline src/test/java/tests/UserTest.java]: get(read)User()

**Review by KirKarpov — CHANGES_REQUESTED** [review]

File: src/test/resources/clientData.properties (hunk 1 above)
Comment (KonstantinLevchukGree) [reply src/test/resources/clientData.properties]: fixed

File: src/main/java/user/User.java (hunk 2 above)
Comment (KonstantinLevchukGree) [reply src/main/java/user/User.java]: fixed

File: src/main/java/user/User.java (hunk 3 above)
Comment (KonstantinLevchukGree) [reply src/main/java/user/User.java]: fixed

File: src/main/java/token/Token.java (hunk 4 above)
Comment (KonstantinLevchukGree) [reply src/main/java/token/Token.java]: fixed

Comment (KonstantinLevchukGree) [reply src/main/java/token/Token.java:11]: fixed

File: src/main/java/utils/authenticationClient/AuthenticationClient.java (hunk 8 above)
Comment (KonstantinLevchukGree) [reply src/main/java/utils/authenticationClient/AuthenticationClient.java]: fixed

File: src/test/java/tests/BaseTest.java (hunk 9 above)
Comment (KonstantinLevchukGree) [reply src/test/java/tests/BaseTest.java]: fixed

File: src/test/java/tests/BaseTest.java (hunk 10 above)
Comment (KonstantinLevchukGree) [reply src/test/java/tests/BaseTest.java]: fixed

File: src/test/java/tests/BaseTest.java (hunk 11 above)
Comment (KonstantinLevchukGree) [reply src/test/java/tests/BaseTest.java]: fixed

File: src/main/java/utils/api/ApiUtils.java (hunk 5 above)
Comment (KonstantinLevchukGree) [reply src/main/java/utils/api/ApiUtils.java]: fixed

File: src/main/java/utils/api/ApiUtils.java (hunk 7 above)
Comment (KonstantinLevchukGree) [reply src/main/java/utils/api/ApiUtils.java]: fixed

File: src/main/java/utils/api/ApiUtils.java (hunk 6 above)
Comment (KonstantinLevchukGree) [reply src/main/java/utils/api/ApiUtils.java]: fixed

File: src/test/resources/clientData.properties (hunk 1 above)
Comment (KonstantinLevchukGree) [reply src/test/resources/clientData.properties]: Does not work

File: src/test/java/tests/UserTest.java (hunk 15 above)
Comment (KonstantinLevchukGree) [reply src/test/java/tests/UserTest.java]: fixed

File: src/test/java/tests/UserTest.java (hunk 14 above)
Comment (KonstantinLevchukGree) [reply src/test/java/tests/UserTest.java]: fixed

File: src/test/java/tests/UserTest.java (hunk 13 above)
Comment (KonstantinLevchukGree) [reply src/test/java/tests/UserTest.java]: fixed

File: src/test/java/tests/UserTest.java (hunk 12 above)
Comment (KonstantinLevchukGree) [reply src/test/java/tests/UserTest.java]: fixed

**Review by KirKarpov — APPROVED** [review]

**Review by KirKarpov — APPROVED** [review]

File: src/main/java/rest/ApplicationClient.java [hunk 16]
```diff
@@ -0,0 +1,96 @@
...
+    public static List<User> getFilterUsers(String sex, String parameter) {
+        CloseableHttpResponse response = CloseableHttpClient.sendGet(applicationData.getProperty("api.users"), parameter, sex);
+        String responseBody = ApiUtils.fromResponseToString(response, HttpStatus.SC_OK).replace("[", "")
+                .replace("]", "").trim();
+        return ApiUtils.getUserFromJson(responseBody);
+    }
+
+    public static void updateUser(User oldUser, User newUser) {
+        String json = "{" + "\"userNewValues\"" + ":" + ApiUtils.fromObjectToJson(newUser)
+                + "," + "\"userToChange\"" + ":" + ApiUtils.fromObjectToJson(oldUser) + "}";
```
Comment (KirKarpov) [inline src/main/java/rest/ApplicationClient.java]: Please create object for your updateUser request.

File: src/main/java/rest/ApplicationClient.java [hunk 17]
```diff
@@ -0,0 +1,96 @@
...
+        CloseableHttpResponse response = CloseableHttpClient.sendPut(applicationData.getProperty("api.users"), json);
+        ApiUtils.fromResponseToString(response, HttpStatus.SC_OK);
```
Comment (KirKarpov) [inline src/main/java/rest/ApplicationClient.java]: This method should reuse updateUser(User oldUser, User newUser, int httpStatus) method

File: src/main/java/rest/ApplicationClient.java [hunk 18]
```diff
@@ -0,0 +1,96 @@
...
+                + "," + "\"userToChange\"" + ":" + ApiUtils.fromObjectToJson(oldUser) + "}";
+        CloseableHttpResponse response = CloseableHttpClient.sendPut(applicationData.getProperty("api.users"), json);
+        ApiUtils.fromResponseToString(response, httpStatus);
+    }
+
+    public static void deleteUser(String json) {
+        CloseableHttpResponse response = CloseableHttpClient.sendDelete(applicationData.getProperty("api.users"), json);
+        String responseBody = ApiUtils.fromResponseToString(response, HttpStatus.SC_NO_CONTENT);
+        ApiUtils.getUserFromJson(responseBody);
+    }
```
Comment (KirKarpov) [inline src/main/java/rest/ApplicationClient.java]: those method should reuse deleteUser(String json, int httpStatus)

File: src/main/java/rest/ApplicationClient.java [hunk 19]
```diff
@@ -0,0 +1,96 @@
...
+    public static void deleteUser(String json, int httpStatus) {
+        CloseableHttpResponse response = CloseableHttpClient.sendDelete(applicationData.getProperty("api.users"), json);
+        String responseBody = ApiUtils.fromResponseToString(response, httpStatus);
+        ApiUtils.getUserFromJson(responseBody);
+    }
+
+    public static String uploadUser(File file) {
+        CloseableHttpResponse response = CloseableHttpClient.sendPost(applicationData.getProperty("api.users.upload"), file);
+        return ApiUtils.fromResponseToString(response, HttpStatus.SC_CREATED).replace("Number of users = ", "").trim();
+    }
```
Comment (KirKarpov) [inline src/main/java/rest/ApplicationClient.java]: method should reuse uploadUser(File file, int httpStatus)

File: src/main/java/tests/DeleteUserTest.java [hunk 20]
```diff
@@ -0,0 +1,50 @@
...
+    }
+
+    @Test
+    public void checkDeleteUserWithRequiredFields() {
+        List<String> availableZipCodes = ApplicationClient.getZipCodes();
+        femaleUser.setZipCode(availableZipCodes.get(0));
+        ApplicationClient.createUser(ApiUtils.fromObjectToJson(femaleUser));
+        femaleUser.setAge(0);
+
+
```
Comment (KirKarpov) [inline src/main/java/tests/DeleteUserTest.java]: It is better logically separate by one line but not two

**Review by KirKarpov — CHANGES_REQUESTED** [review]

File: src/main/java/tests/DeleteUserTest.java (hunk 20 above)
Comment (KonstantinLevchukGree) [reply src/main/java/tests/DeleteUserTest.java]: fixed

File: src/main/java/rest/ApplicationClient.java (hunk 19 above)
Comment (KonstantinLevchukGree) [reply src/main/java/rest/ApplicationClient.java]: fixed

File: src/main/java/rest/ApplicationClient.java (hunk 18 above)
Comment (KonstantinLevchukGree) [reply src/main/java/rest/ApplicationClient.java]: fixed

File: src/main/java/rest/ApplicationClient.java (hunk 17 above)
Comment (KonstantinLevchukGree) [reply src/main/java/rest/ApplicationClient.java]: fixed

File: src/main/java/rest/ApplicationClient.java (hunk 16 above)
Comment (KonstantinLevchukGree) [reply src/main/java/rest/ApplicationClient.java]: fixed

**Review by KirKarpov — APPROVED** [review]

**Pull request merged**